### PR TITLE
[#9] Implemented Logger class.

### PR DESCRIPTION
**Pull request by tamerharsh**

### CODE REVIEW

File: include/logger/common.hpp [hunk 1]
```diff
@@ -22,5 +22,15 @@ struct ExtractedLogCredentials
...
 };
+
+/**
+ * @brief  Data Structure to user specified log related details.
+ */
+struct LogSettings
+{
+ logger::LEVEL desired_level=logger::LEVEL::TRACE;
+ std::string logger_name="default logger";
+
```
Comment (inbangsa) [inline include/logger/common.hpp]: remove space

File: include/logger/err.hpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ *@brief namespace for logger lib .
+ */
+namespace logger {
+
+/*
+ *@brief namespace for error code .
+ */
+namespace error_code {
+
+  const std::string ERROR_101{ "[Error 101] : Data found to be empty." };
```
Comment (inbangsa) [inline include/logger/err.hpp]: error string cant be error code 

Check this, Try to do it like this.
https://github.com/ifm/ifm3d/blob/master/modules/camera/src/libifm3d_camera/err.cpp

File: include/logger/logger_lib.hpp [hunk 3]
```diff
@@ -1,9 +1,302 @@
...
 
+/*
+ *@brief To store logger name and desired severity level.
+ */
+auto log_settings = std::make_shared<logger::LogSettings>();
+
+/*
+ *@brief To record the data to assiged sink.
+ */
+std::shared_ptr<logger::ISink> sink_ptr = std::make_shared<logger::DefaultSink>();
```
Comment (inbangsa) [inline include/logger/logger_lib.hpp]: Why Sink we can create logger object directly. A better approch will be if user create a logger class we used that Or if they directly we create a default logger object for logging. 

Rest code is good , Try to create a log object and take away all globa functions

**Review by inbangsa — CHANGES_REQUESTED** [review]

File: include/logger/err.hpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+/*
+ *@brief namespace for logger lib .
+ */
+namespace logger {
+
+/*
+ *@brief namespace for error code .
+ */
+namespace error_code {
```
Comment (sachinsnale) [inline include/logger/err.hpp]: what sort of error's are you expecting. how will be the usage.

File: include/logger/logger_lib.hpp [hunk 5]
```diff
@@ -1,9 +1,357 @@
...
+ *@param logger_name logger name
+ */
+#define SET_LOGGER_NAME(logger_name) logger::set_logger_name(logger_name)
 namespace logger {
+/*
+ *@brief Sets the log severity level.
+ *@param level  desired log severity level
+ *@retval None
+ */
+void set_severity_level(logger::LEVEL level);
```
Comment (inbangsa) [inline include/logger/logger_lib.hpp]: can we take away All following global function if we inject logger inside macro ? 

I would like you to give a thought on this part of the code. All function are repeated  twice so take away the redundant code.

**Review by inbangsa — DISMISSED** [review]

File: include/logger/logger_lib.hpp [hunk 6]
```diff
@@ -1,9 +1,357 @@
...
+   *@brief A Pointer to the struct ExtractedLogCredentials for storing the extracted log data such as date ,time,
+   *filename etc.
+   */
+  std::shared_ptr<logger::ExtractedLogCredentials> log_credentials;
+};
+
+/*
+ *@brief A Logger class object to be used internally to do various operations for macros.
+ */
+Logger internal_object;
```
Comment (inbangsa) [inline include/logger/logger_lib.hpp]: a global Logger OR a static logger ?

File: include/logger/err.hpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+     *@retval None
+     */
+    LoggerException(int error_code);
+
+    /*
+     *@brief Gives the error message whenever error is catched.
+     *@param None
+     *@retval const char*
+     */
+    virtual const char *what() const noexcept;
```
Comment (sachinsnale) [inline include/logger/err.hpp]: This is virtual function in the base, so you don't need to write virtual before function here again.
just add override keyword at the end of the function.

File: include/logger/err.hpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+     *@brief Variable to store error code.
+     */
+    int error_code;
+  };
+
+};// namespace error_handling
+
+logger::error_handling::LoggerException::LoggerException(int error_code) : std::exception(), error_code(error_code) {}
+const char *error_handling::LoggerException::what() const noexcept { return GetErrorMessage(this->get_errro_code()); }
+int error_handling::LoggerException::get_errro_code() const noexcept { return this->error_code; }
```
Comment (sachinsnale) [inline include/logger/err.hpp]: either do the inline implementation or move them to '.cpp' file.

File: include/logger/common.hpp (hunk 1 above)
Comment (sachinsnale) [reply include/logger/common.hpp]: along with this please check all over the places, where unnecessary line spaces are used.

File: include/logger/err.hpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+  /*
+   *@brief Gives error message for particular error type.
+   *@param error_code error code
+   *@retval const char*
+   */
+  const char *GetErrorMessage(const int error_code)
+  {
+    switch (error_code) {
+
```
Comment (sachinsnale) [inline include/logger/err.hpp]: no need of line space

**Review by sachinsnale — CHANGES_REQUESTED** [review]
i see that you keep everything in hpp, is there any reason for this.
as cpp user its good practice to separate them in hpp and cpp, as an when possible, specially for classes

File: include/logger/err.hpp (hunk 8 above)
Comment (tamerharsh) [reply include/logger/err.hpp]: Actually we need header only lib. The intent is  that user need to include the  .hpp files in include folder and is good to go. 

Almost all such libs like fmt , spdlog, JSON etc  supports  header only lib  .

File: include/logger/err.hpp (hunk 9 above)
Comment (tamerharsh) [reply include/logger/err.hpp]: Actually i use clang format , still i will check why space issues are there.

File: include/logger/logger_lib.hpp [hunk 10]
```diff
@@ -134,95 +19,99 @@ class Logger
...
+    const std::string& logger_name = "runtime_default_logger",
+    logger::log_level log_level = logger::log_level::TRACE);
+
+  //Copy and Copy Assignment Operator set to default.
+  Logger(const Logger&) = default;
+  Logger& operator=(const Logger&) = default;
 
+  //Move and Move Assignment Operator set to default.
+  Logger(Logger&&)=default;
+  Logger& operator=(Logger&&) = default;
```
Comment (sachinsnale) [inline include/logger/logger_lib.hpp]: Anyway if you don't have your own implementation, compiler will take default.
not necessary to add this code.

**Review by sachinsnale — APPROVED** [review]

**Comment by tamerharsh** [comment]
Merging this pull request  with approval .

**Comment by tamerharsh** [comment]
Merging this pull request with approval.

**Comment by tamerharsh** [comment]
Closing PR with approval.

**Comment by tamerharsh** [comment]
Closing PR with approval.

**Comment by tamerharsh** [comment]
Closing PR.

**Comment by tamerharsh** [comment]
Merging to master .

**Comment by tamerharsh** [comment]
Merging  with master.

**Pull request closed without merging**